tensornetwork/contractors/stochastic_contractor.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2019 The TensorNetwork Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Stochastic Network Contraction."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import random
+import numpy as np
+from tensornetwork import tensornetwork
+
+
+def get_random_edge(network: tensornetwork.TensorNetwork) -> tensornetwork.Edge:
+  """Returns a random connected edge.
+  If it finds a trace edge it contracts it.
+  Args:
+    network: Connected TensorNetwork we are trying to contract.
+  Returns:
+    edge: A random connected edge from the network.
+  """
+  node = random.choice(tuple(network.nodes_set))
+  edge = random.choice(node.edges)
+  while edge.is_dangling():
+    edge = random.choice(node.edges)
+  if edge.node1 is edge.node2:
+    # Contract the trace edge and look for a new one
+    network.contract(edge)
+    return get_random_edge(network)
+  return edge
+
+
+def edge_cost(edge: tensornetwork.Edge) -> int:
+  """Calculates cost of contracting an edge.
+  If A and B are the tensors that share the given `edge`, cost is defined as:
+  cost = dims(A * B) - max(dims(A), dims(B)), where
+  * denotes contraction of all shared edges (`contract_between`) and
+  dims(X) is the total dimension of tensor X (product of sizes of all axes).
+  Args:
+    edge: Edge to calculate the cost.
+  Returns:
+    cost: Cost of the given edge.
+  """
+  # TODO: Verify whether this loss makes sense.
+  node1, node2 = edge.node1, edge.node2
+  # Calculate dimension of all shared edges
+  nodes = {node1, node2}
+  shared_dimension = 1
+  for edge in node1.edges:
+    if set(edge.get_nodes()) == nodes:
+      shared_dimension *= int(edge.node1.get_tensor().shape[edge.axis1])
+
+  dimension1 = int(np.prod(node1.get_tensor().shape))
```

Comment:
You should try to gracefully handle tensors with None shape. Either raise an error of change the calculation, your choice.

Reply:
In the last update I just ignore None sizes in the cost calculation. I am not sure if that's a good choice but probably better than non supporting this at all. The contractor is mostly random so it won't be perfect anyway.